models/base/document.py

```diff
@@ -0,0 +1,231 @@
+        :return: the first referrer matching the criteria
+        """
+        for referrer in self._referenced_by:
+            if referrer_type is not None and not isinstance(referrer, referrer_type):
+                continue
+            if field_name is not None and not self in getattr(referrer, field_name):
+                continue
+            return referrer
+        raise self.ReferrerNotFound(referrer_type, field_name)
+
+    def __eq__(self, other):
+        return self._data == other._data
+
+    def _add_referrer(self, referer):
+        if referer not in self._referenced_by:
+            self._referenced_by.append(referer)
+
+    def _remove_referrer(self, referer):
+        if referer in self._referenced_by:
+            self._referenced_by.remove(referer)
+
+    def __unlink_referee(self, referee):
+        for field_name, field in self._fields.items():
+            if isinstance(field, ReferenceDocumentsField):
+                documents = getattr(self, field_name)
+                if referee in documents:
+                    documents.remove(referee)
+        self.save()
+
+    def __str__(self):
+        return f'{self.__class__.__name__}({self._data})'
+
+
+class IndexedDocument(Document, metaclass=MetaIndexedDocument):
+    """
+    Base class for all root level documents, directly persisted to disk.
+    A primary key must be defined to index all active documents.
+    Documents are persisted as the index to all documents in this class.
+    The default persistence path is data/{classname}
+    to change the path, override the _persistence_path property.
+    """
+
+    def __init__(self, **kwargs):
+        if not self._primary_key:
+            raise Document.PrimaryKeyNotDefinedError(self)
+        super().__init__(**kwargs)
+
+    @classmethod
+    def reload(cls) -> None:
+        """
+        Reload the index from disk.
+        """
+        try:
+            with open(cls._persistence_path, 'rb') as f:
+                cls.__objects = pickle.load(f)
+        except FileNotFoundError:
+            cls.__objects = {}
+
+    def save(self) -> None:
+        """
+        Save all documents of the same type (i.e. the index) to disk.
+        Also save all root-level documents referencing this document.
+        """
+        self.__class__.__objects[self.key] = self
+        os.makedirs(os.path.dirname(self._persistence_path), exist_ok=True)
+        with open(self._persistence_path, 'wb') as f:
+            pickle.dump(self.__class__.__objects, f)
+        super().save()
+
+    @classmethod
+    def find(cls, key) -> Union[None, IndexedDocument]:
+        """
+        Find a document by its primary key.
+        :param key: the primary key value
+        :return: the document, or None if not found
+        """
+        return cls.__objects.get(key)
+
+    @classmethod
+    def all(cls) -> list[IndexedDocument]:
+        """
+        Get all documents of this type.
```

Comment:
What does "documents of this type" mean? 
What are the different "types" of documents?

Reply:
It means the class. 
For example, when you subclass from `IndexedDocument`, the `all` method return all the instances of your class.